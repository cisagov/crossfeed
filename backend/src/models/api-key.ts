import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToOne
} from 'typeorm';
import { User } from './';

@Entity()
export class ApiKey extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @ManyToOne((type) => User, (user) => user.apiKeys, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  user: User;

  @Column({
    type: 'timestamp',
    nullable: true
  })
  lastUsed: Date | null;

  @Column({
    // length based on hex formatted sha256
    length: 64,
    type: 'varchar'
  })
  hashedKey: string;

  @Column({
    length: 4,
    type: 'varchar'
  })
  lastFour: string;
}
