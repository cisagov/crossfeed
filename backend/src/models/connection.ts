import { createConnection, Connection } from 'typeorm';
import {
  Domain,
  Service,
  Vulnerability,
  Scan,
  Organization,
  User,
  Role,
  ScanTask
} from '.';

let connection: Connection | null = null;

const connectDb = async (logging?: boolean) => {
  const connection = createConnection({
    type: 'postgres',
    host: process.env.DB_HOST,
    port: parseInt(process.env.DB_PORT ?? ''),
    username: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_NAME,
    entities: [
      Domain,
      Service,
      Vulnerability,
      Scan,
      Organization,
      User,
      Role,
      ScanTask
    ],
    synchronize: false,
    name: 'default',
    dropSchema: false,
    logging: logging ?? false
  });
  return connection;
};

export const connectToDatabase = async (logging?: boolean) => {
  if (!connection?.isConnected) {
    connection = await connectDb(logging);
  }
  if (typeof jest === 'undefined') {
    console.log('=> DB Connected');
  }
  return connection;
};
