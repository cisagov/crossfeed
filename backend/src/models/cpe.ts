import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToMany,
  BaseEntity,
  Unique
} from 'typeorm';
import { Cve } from './cve';

@Entity()
@Unique(['name', 'version', 'vendor'])
export class Cpe extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  name: string;

  @Column()
  version: string;

  @Column()
  vendor: string;

  @Column()
  lastSeenAt: Date;

  @ManyToMany(() => Cve, (cve) => cve.cpes)
  cves: Cve[];
}
