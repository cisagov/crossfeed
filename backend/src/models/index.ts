export * from './domain';
export * from './service';
export * from './connection';
export * from './vulnerability';
export * from './scan';
export * from './organization';
export * from './user';
export * from './role';
export * from './scan-task';
