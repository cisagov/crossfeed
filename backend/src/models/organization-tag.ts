import {
  Entity,
  Index,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToMany,
  JoinTable,
  Column
} from 'typeorm';
import { Organization, Scan } from '.';

@Entity()
@Index(['name'], { unique: true })
export class OrganizationTag extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column()
  name: string;

  /**
   * Organizations that are labeled with this tag
   */
  @ManyToMany((type) => Organization, (organization) => organization.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  @JoinTable()
  organizations: Organization[];

  /**
   * Scans that have this tag enabled, and will run against all tagged organizations
   */
  @ManyToMany((type) => Scan, (scan) => scan.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  scans: Scan[];
}
