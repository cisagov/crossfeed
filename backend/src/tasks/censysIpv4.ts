import { Domain, Service } from '../models';
import { plainToClass } from 'class-transformer';
import saveDomainsToDb from './helpers/saveDomainsToDb';
import { CommandOptions } from './ecs-client';
import { CensysIpv4Data } from 'src/models/generated/censysIpv4';
import { mapping } from './censys/mapping';
import saveServicesToDb from './helpers/saveServicesToDb';
import getAllDomains from './helpers/getAllDomains';
import * as zlib from 'zlib';
import * as readline from 'readline';
import got from 'got';
import PQueue from 'p-queue';
import pRetry from 'p-retry';
import axios from 'axios';

interface IpToDomainsMap {
  [ip: string]: Domain[];
}

const auth = {
  username: process.env.CENSYS_API_ID!,
  password: process.env.CENSYS_API_SECRET!
};

const CENSYS_IPV4_ENDPOINT = 'https://censys.io/api/v1/data/ipv4_2018/';

// Sometimes, a field might contain null characters, but we can't store null
// characters in a string field in PostgreSQL. For example, a site might have
// a banner ending with "</body>\r\n</html>\u0000" or "\\u0000".
const sanitizeStringField = (input) =>
  input.replace(/\\u0000/g, '').replace(/\0/g, '');

const downloadPath = async (
  path: string,
  ipToDomainsMap: IpToDomainsMap,
  i: number,
  numFiles: number,
  commandOptions: CommandOptions
): Promise<void> => {
  console.log(`i: ${i} of ${numFiles}: starting download of url ${path}`);

  const domains: Domain[] = [];
  const services: Service[] = [];
  const gunzip = zlib.createGunzip();
  const downloadStream = got.stream.get(path, { ...auth });
  const unzippedStream = downloadStream.pipe(gunzip);
  // readInterface lets us stream the JSON file line-by-line
  const readInterface = readline.createInterface({
    input: unzippedStream
  });
  await new Promise((resolve, reject) => {
    downloadStream.on('error', reject);
    readInterface.on('line', function (line) {
      const item: CensysIpv4Data = JSON.parse(line);

      let matchingDomains = ipToDomainsMap[item.ip!] || [];
      if (process.env.IS_LOCAL && typeof jest === 'undefined') {
        // For local development: just randomly match domains
        // (this behavior is not present when running tests
        // through jest, though).
        // eslint-disable-next-line prefer-spread
        matchingDomains = [].concat
          .apply([], Object.values(ipToDomainsMap)) // get a list of all domains in the domain map
          .filter(() => Math.random() < 0.00001);
      }
      for (const matchingDomain of matchingDomains) {
        domains.push(
          plainToClass(Domain, {
            name: matchingDomain.name,
            organization: matchingDomain.organization,
            asn: item.autonomous_system?.asn,
            ip: item.ip,
            country: item.location?.country_code
          })
        );
        for (const key in item) {
          if (key.startsWith('p') && mapping[key]) {
            const service = Object.keys(item[key] as any)[0];
            const s = {
              ...mapping[key](item[key]),
              service,
              discoveredBy: { id: commandOptions.scanId },
              port: Number(key.slice(1)),
              domain: matchingDomain,
              lastSeen: new Date(Date.now()),
              censysIpv4Results: JSON.parse(
                sanitizeStringField(JSON.stringify(item[key]))
              )
            };
            for (const k in s) {
              if (typeof s[k] === 'string') {
                s[k] = sanitizeStringField(s[k]);
              }
            }
            services.push(plainToClass(Service, s));
          }
        }
      }
    });
    readInterface.on('close', resolve);
    readInterface.on('SIGINT', reject);
    readInterface.on('SIGCONT', reject);
    readInterface.on('SIGTSTP', reject);
  });
  if (!domains.length) {
    console.log(
      `censysipv4 - processed file ${i} of ${numFiles}: got no results`
    );
  } else {
    console.log(
      `censysipv4 - processed file ${i} of ${numFiles}: got some results: ${domains.length} domains and ${services.length} services`
    );
  }

  await saveDomainsToDb(domains);
  await saveServicesToDb(services);
};

export const handler = async (commandOptions: CommandOptions) => {
  const { chunkNumber, numChunks } = commandOptions;

  if (chunkNumber === undefined || numChunks === undefined) {
    throw new Error('Chunks not specified.');
  }

  const {
    data: { results }
  } = await axios.get(CENSYS_IPV4_ENDPOINT, { auth });

  const {
    data: { files }
  } = await axios.get(results.latest.details_url, { auth });

  const allDomains = await getAllDomains();

  const queue = new PQueue({ concurrency: 5 });

  const numFiles = Object.keys(files).length;
  const fileNames = Object.keys(files).sort();
  const jobs: Promise<void>[] = [];

  let startIndex = Math.floor(((1.0 * chunkNumber) / numChunks) * numFiles);
  let endIndex =
    Math.floor(((1.0 * (chunkNumber + 1)) / numChunks) * numFiles) - 1;

  if (process.env.IS_LOCAL && typeof jest === 'undefined') {
    // For local testing.
    startIndex = 0;
    endIndex = 1;
  }

  const ipToDomainsMap: IpToDomainsMap = allDomains.reduce<IpToDomainsMap>(
    (map: IpToDomainsMap, domain: Domain) => {
      if (!map[domain.ip]) {
        map[domain.ip] = [];
      }
      map[domain.ip].push(domain);
      return map;
    },
    {}
  );

  for (let i = startIndex; i <= endIndex; i++) {
    const idx = i;
    const fileName = fileNames[idx];
    jobs.push(
      queue.add(() =>
        pRetry(
          () =>
            downloadPath(
              files[fileName].download_path,
              ipToDomainsMap,
              idx,
              numFiles,
              commandOptions
            ),
          {
            // Perform less retries on jest to make tests faster
            retries: typeof jest === 'undefined' ? 5 : 2,
            randomize: true
          }
        )
      )
    );
  }
  console.log(`censysipv4: scheduled all tasks`);
  await Promise.all(jobs);

  console.log(`censysipv4 done`);
};
