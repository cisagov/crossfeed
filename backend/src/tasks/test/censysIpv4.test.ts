import { handler as censysIpv4 } from '../censysIpv4';
import * as zlib from 'zlib';
import * as nock from 'nock';
import * as path from 'path';
import * as fs from 'fs';
import { connectToDatabase, Domain, Organization, Scan } from '../../models';

const RealDate = Date;

const authHeaders = {
  reqheaders: {
    Authorization:
      'Basic ' +
      Buffer.from(
        `${process.env.CENSYS_API_ID}:${process.env.CENSYS_API_SECRET}`
      ).toString('base64')
  }
};

describe('censys ipv4', () => {
  let organization;
  let scan;
  beforeEach(async () => {
    await connectToDatabase();
    global.Date.now = jest.fn(() => new Date('2019-04-22T10:20:30Z').getTime());
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
    scan = await Scan.create({
      name: 'censysIpv4',
      arguments: {},
      frequency: 999
    }).save();
    await Domain.create({
      name: 'first_file_testdomain1',
      ip: '153.126.148.60',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain2',
      ip: '31.134.10.156',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain3',
      ip: '153.126.148.61',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain4',
      ip: '85.24.146.152',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain5',
      ip: '45.79.207.117',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain6',
      ip: '156.249.159.119',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain7',
      ip: '221.10.15.220',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain8',
      ip: '81.141.166.145',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain9',
      ip: '24.65.82.187',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain10',
      ip: '52.74.149.117',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain11',
      ip: '31.134.10.156',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain12',
      ip: '1.1.1.1',
      organization
    }).save();
  });

  afterEach(async () => {
    global.Date = RealDate;
  });

  const checkDomains = async (organization) => {
    const domains = await Domain.find({
      where: { organization },
      relations: ['organization', 'services']
    });
    expect(
      domains
        .sort((a, b) => a.name.localeCompare(b.name))
        .map((e) => ({
          ...e,
          services: e.services.map((s) => ({
            ...s,
            id: null,
            updatedAt: null,
            createdAt: null
          })),
          organization: null,
          id: null,
          updatedAt: null,
          createdAt: null,
          syncedAt: null
        }))
    ).toMatchSnapshot();
    expect(domains.filter((e) => !e.organization).length).toEqual(0);
  };

  test('basic test', async () => {
    nock('https://censys.io', authHeaders)
      .get('/api/v1/data/ipv4_2018/')
      .reply(200, {
        results: {
          latest: {
            details_url: 'https://censys.io/api/v1/data/ipv4_2018/20200719'
          }
        }
      });
    nock('https://censys.io', authHeaders)
      .get('/api/v1/data/ipv4_2018/20200719')
      .reply(200, {
        files: {
          'first_file.json.gz': {
            file_type: 'json',
            compressed_size: 2568162,
            compression_type: 'gzip',
            compressed_md5_fingerprint: '334183d35efade2336033b38c4c528a6',
            download_path:
              'https://data-01.censys.io/snapshots/ipv4/20200719/first_file.json.gz'
          },
          'second_file.json.gz': {
            file_type: 'json',
            compressed_size: 2568162,
            compression_type: 'gzip',
            compressed_md5_fingerprint: '334183d35efade2336033b38c4c528a6',
            download_path:
              'https://data-01.censys.io/snapshots/ipv4/20200719/second_file.json.gz'
          }
        }
      });
    const firstFileContents = fs.readFileSync(
      path.join(__dirname, '../helpers/__mocks__/censysIpv4Sample.json')
    );
    const secondFileContents = JSON.stringify({
      address: '1.1.1.1',
      ipint: '1319753985',
      updated_at: '2020-07-05T15:41:45Z',
      ip: '1.1.1.1',
      location: {
        country_code: 'TR',
        continent: 'Asia',
        city: 'Akalan',
        postal_code: '55400',
        timezone: 'Europe/Istanbul',
        province: 'Samsun',
        latitude: 41.2748,
        longitude: 35.7197,
        registered_country: 'Turkey',
        registered_country_code: 'TR',
        country: 'Turkey'
      },
      autonomous_system: {
        description: 'TTNET',
        routed_prefix: '78.169.128.0/17',
        asn: '9121',
        country_code: 'TR',
        name: 'TTNET',
        path: ['7018', '3320', '9121']
      },
      ports: ['53', '8080'],
      protocols: ['53/dns', '8080/http'],
      ipinteger: '31041870',
      version: '0',
      p53: {
        dns: {
          lookup: {
            errors: false,
            open_resolver: false,
            questions: [{ name: 'c.afekv.com', type: 'A' }],
            resolves_correctly: false,
            support: true,
            timestamp: '2020-07-05T15:41:45Z'
          }
        }
      },
      p8080: {
        http: {
          get: {
            body: `these characters should not show up in the snapshot:
              null: \u0000
              another null: \0
              another null: \\u0000
              `
          }
        }
      },
      tags: ['dns']
    });

    nock('https://data-01.censys.io', authHeaders)
      .persist()
      .get('/snapshots/ipv4/20200719/first_file.json.gz')
      .reply(200, zlib.gzipSync(firstFileContents))
      .get('/snapshots/ipv4/20200719/second_file.json.gz')
      .reply(200, zlib.gzipSync(secondFileContents));
    jest.setTimeout(30000);
    await censysIpv4({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId',
      chunkNumber: 0,
      numChunks: 1
    });

    await checkDomains(organization);
  });

  test('http failure should retry', async () => {
    nock('https://censys.io', authHeaders)
      .get('/api/v1/data/ipv4_2018/')
      .reply(200, {
        results: {
          latest: {
            details_url: 'https://censys.io/api/v1/data/ipv4_2018/20200719'
          }
        }
      });
    nock('https://censys.io', authHeaders)
      .get('/api/v1/data/ipv4_2018/20200719')
      .reply(200, {
        files: {
          'failed_file.json.gz': {
            file_type: 'json',
            compressed_size: 2568162,
            compression_type: 'gzip',
            compressed_md5_fingerprint: '334183d35efade2336033b38c4c528a6',
            download_path:
              'https://data-01.censys.io/snapshots/ipv4/20200719/failed_file.json.gz'
          }
        }
      });

    nock('https://data-01.censys.io')
      .get('/snapshots/ipv4/20200719/failed_file.json.gz')
      .reply(429, 'too many requests');

    nock('https://data-01.censys.io')
      .get('/snapshots/ipv4/20200719/failed_file.json.gz')
      .reply(200, zlib.gzipSync(JSON.stringify({})));

    jest.setTimeout(30000);
    await censysIpv4({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId',
      chunkNumber: 0,
      numChunks: 1
    });

    await checkDomains(organization);
  });

  test('repeated http failures should throw an error', async () => {
    nock('https://censys.io', authHeaders)
      .get('/api/v1/data/ipv4_2018/')
      .reply(200, {
        results: {
          latest: {
            details_url: 'https://censys.io/api/v1/data/ipv4_2018/20200719'
          }
        }
      });
    nock('https://censys.io', authHeaders)
      .get('/api/v1/data/ipv4_2018/20200719')
      .reply(200, {
        files: {
          'failed_file_2.json.gz': {
            file_type: 'json',
            compressed_size: 2568162,
            compression_type: 'gzip',
            compressed_md5_fingerprint: '334183d35efade2336033b38c4c528a6',
            download_path:
              'https://data-01.censys.io/snapshots/ipv4/20200719/failed_file_2.json.gz'
          }
        }
      });

    nock('https://data-01.censys.io')
      .persist()
      .get('/snapshots/ipv4/20200719/failed_file_2.json.gz')
      .reply(429, 'too many requests');

    jest.setTimeout(30000);
    await expect(
      censysIpv4({
        organizationId: organization.id,
        organizationName: 'organizationName',
        scanId: scan.id,
        scanName: 'scanName',
        scanTaskId: 'scanTaskId',
        chunkNumber: 0,
        numChunks: 1
      })
    ).rejects.toThrow('Response code 429');

    await checkDomains(organization);
  });
});
