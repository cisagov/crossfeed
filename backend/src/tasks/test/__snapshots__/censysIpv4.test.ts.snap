// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`censys ipv4 basic test 1`] = `
Array [
  Object {
    "asn": "7684",
    "cloudHosted": false,
    "country": "JP",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "153.126.148.60",
    "name": "first_file_testdomain1",
    "organization": null,
    "reverseName": "first_file_testdomain1",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "16509",
    "cloudHosted": false,
    "country": "SG",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "52.74.149.117",
    "name": "first_file_testdomain10",
    "organization": null,
    "reverseName": "first_file_testdomain10",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "48347",
    "cloudHosted": false,
    "country": "RU",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "31.134.10.156",
    "name": "first_file_testdomain11",
    "organization": null,
    "reverseName": "first_file_testdomain11",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "9121",
    "cloudHosted": false,
    "country": "TR",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "1.1.1.1",
    "name": "first_file_testdomain12",
    "organization": null,
    "reverseName": "first_file_testdomain12",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "48347",
    "cloudHosted": false,
    "country": "RU",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "31.134.10.156",
    "name": "first_file_testdomain2",
    "organization": null,
    "reverseName": "first_file_testdomain2",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "7684",
    "cloudHosted": false,
    "country": "JP",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "153.126.148.60",
    "name": "first_file_testdomain3",
    "organization": null,
    "reverseName": "first_file_testdomain3",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "8473",
    "cloudHosted": false,
    "country": "SE",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "85.24.146.152",
    "name": "first_file_testdomain4",
    "organization": null,
    "reverseName": "first_file_testdomain4",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "63949",
    "cloudHosted": false,
    "country": "US",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "45.79.207.117",
    "name": "first_file_testdomain5",
    "organization": null,
    "reverseName": "first_file_testdomain5",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "26484",
    "cloudHosted": false,
    "country": "US",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "156.249.159.119",
    "name": "first_file_testdomain6",
    "organization": null,
    "reverseName": "first_file_testdomain6",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "221.10.15.220",
    "name": "first_file_testdomain7",
    "organization": null,
    "reverseName": "first_file_testdomain7",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "2856",
    "cloudHosted": false,
    "country": "GB",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "81.141.166.145",
    "name": "first_file_testdomain8",
    "organization": null,
    "reverseName": "first_file_testdomain8",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": "6327",
    "cloudHosted": false,
    "country": "CA",
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "24.65.82.187",
    "name": "first_file_testdomain9",
    "organization": null,
    "reverseName": "first_file_testdomain9",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
]
`;

exports[`censys ipv4 http failure should retry 1`] = `
Array [
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "153.126.148.60",
    "name": "first_file_testdomain1",
    "organization": null,
    "reverseName": "first_file_testdomain1",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "52.74.149.117",
    "name": "first_file_testdomain10",
    "organization": null,
    "reverseName": "first_file_testdomain10",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "31.134.10.156",
    "name": "first_file_testdomain11",
    "organization": null,
    "reverseName": "first_file_testdomain11",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "1.1.1.1",
    "name": "first_file_testdomain12",
    "organization": null,
    "reverseName": "first_file_testdomain12",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "31.134.10.156",
    "name": "first_file_testdomain2",
    "organization": null,
    "reverseName": "first_file_testdomain2",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "153.126.148.60",
    "name": "first_file_testdomain3",
    "organization": null,
    "reverseName": "first_file_testdomain3",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "85.24.146.152",
    "name": "first_file_testdomain4",
    "organization": null,
    "reverseName": "first_file_testdomain4",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "45.79.207.117",
    "name": "first_file_testdomain5",
    "organization": null,
    "reverseName": "first_file_testdomain5",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "156.249.159.119",
    "name": "first_file_testdomain6",
    "organization": null,
    "reverseName": "first_file_testdomain6",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "221.10.15.220",
    "name": "first_file_testdomain7",
    "organization": null,
    "reverseName": "first_file_testdomain7",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "81.141.166.145",
    "name": "first_file_testdomain8",
    "organization": null,
    "reverseName": "first_file_testdomain8",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "24.65.82.187",
    "name": "first_file_testdomain9",
    "organization": null,
    "reverseName": "first_file_testdomain9",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
]
`;

exports[`censys ipv4 repeated http failures should throw an error 1`] = `
Array [
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "153.126.148.60",
    "name": "first_file_testdomain1",
    "organization": null,
    "reverseName": "first_file_testdomain1",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "52.74.149.117",
    "name": "first_file_testdomain10",
    "organization": null,
    "reverseName": "first_file_testdomain10",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "31.134.10.156",
    "name": "first_file_testdomain11",
    "organization": null,
    "reverseName": "first_file_testdomain11",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "1.1.1.1",
    "name": "first_file_testdomain12",
    "organization": null,
    "reverseName": "first_file_testdomain12",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "31.134.10.156",
    "name": "first_file_testdomain2",
    "organization": null,
    "reverseName": "first_file_testdomain2",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "153.126.148.60",
    "name": "first_file_testdomain3",
    "organization": null,
    "reverseName": "first_file_testdomain3",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "85.24.146.152",
    "name": "first_file_testdomain4",
    "organization": null,
    "reverseName": "first_file_testdomain4",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "45.79.207.117",
    "name": "first_file_testdomain5",
    "organization": null,
    "reverseName": "first_file_testdomain5",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "156.249.159.119",
    "name": "first_file_testdomain6",
    "organization": null,
    "reverseName": "first_file_testdomain6",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "221.10.15.220",
    "name": "first_file_testdomain7",
    "organization": null,
    "reverseName": "first_file_testdomain7",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "81.141.166.145",
    "name": "first_file_testdomain8",
    "organization": null,
    "reverseName": "first_file_testdomain8",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "24.65.82.187",
    "name": "first_file_testdomain9",
    "organization": null,
    "reverseName": "first_file_testdomain9",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
]
`;
